File: algorithms/1_easy/C/04006_count_valid_prefixes/solution.c
typedef ptrdiff_t isize;

int countValidPrefixes(const char *restrict s) {
    size_t cnt = 0;
    isize balance = 0;
    for (const char *it = s; *it != '\0'; it += 1) {
        switch (*it) {
            case '0':
                balance -= 1;
                break;
            case '1':
                balance += 1;
                break;
            default:
                return -1;
        }

        if (ABS(balance) <= 1) { cnt += 1; }
    }

    return (int)cnt;
}

File: algorithms/1_easy/C/04006_count_valid_prefixes/solution.rs
impl Solution {
    pub fn count_valid_prefixes(s: String) -> i32 {
        let mut cnt: usize = 0;
        let mut balance: isize = 0;
        for &b in s.as_bytes() {
            match b {
                b'0' => balance -= 1,
                b'1' => balance += 1,
                _ => unreachable!(),
            }

            if balance.abs() <= 1 { cnt += 1; }
        }

        cnt.try_into().unwrap()
    }
}
